components/TARGET_PSA/services/crypto/COMPONENT_SPE/psa_crypto_partition.c

```diff
@@ -134,24 +147,48 @@ static void psa_mac_operation(void)
                 }
 
                 case PSA_MAC_UPDATE: {
-                    uint8_t *input_ptr = mbedtls_calloc(1, msg.in_size[1]);
-                    if (input_ptr == NULL) {
-                        status = PSA_ERROR_INSUFFICIENT_MEMORY;
-                        break;
-                    }
 
-                    bytes_read = psa_read(msg.handle, 1, input_ptr,
-                                          msg.in_size[1]);
+                    size_t data_remaining = msg.in_size[1];
+                    size_t size_to_read = 0;
+                    psa_status_t failure_status = PSA_SUCCESS;
 
-                    if (bytes_read != msg.in_size[1]) {
-                        SPM_PANIC("SPM read length mismatch");
+                    if (input_buffer_ptr == NULL)
+                    {
+                        input_buffer_ptr = mbedtls_calloc(1, MAX_DATA_CHUNK_SIZE);
+                        if (input_buffer_ptr == NULL) {
+                            status = PSA_ERROR_INSUFFICIENT_MEMORY;
+                            break;
+                        }
                     }
 
-                    status = psa_mac_update(msg.rhandle,
-                                            input_ptr,
-                                            msg.in_size[1]);
+                    while (data_remaining > 0)
+                    {
+                        size_to_read = data_remaining;
+                        if (size_to_read > MAX_DATA_CHUNK_SIZE)
+                        {
+                            size_to_read = MAX_DATA_CHUNK_SIZE;
+                        }
+                        bytes_read = psa_read(msg.handle, 1, input_buffer_ptr,
+                                              size_to_read);
+
+                        if (bytes_read != size_to_read) {
+                            SPM_PANIC("SPM read length mismatch");
+                        }
+
+                        status = psa_mac_update(msg.rhandle,
+                                                input_buffer_ptr,
+                                                bytes_read);
+                        if (status != PSA_SUCCESS)
```

Comment:
In case multiple calls to `psa_xxx_update` fail for different reasons (returning different error codes) the **last** error code is recorded in `failure_status` and returned to the caller, wouldn't it make more sense to record the first error and return the root failure?
```suggestion
                        if (status != PSA_SUCCESS && failure_status == PSA_SUCCESS)
```

Reply:
ok, sounds reasonable, will return the first error instead of the last.